gcloud/monitoring/client.py

```diff
@@ -0,0 +1,189 @@
+
+        :type end_time: :class:`datetime.datetime` or None
+        :param end_time: The end time (inclusive) of the time interval
+            for which results should be returned, as a datetime object.
+            The default is the start of the current minute.
+
+            The start time (exclusive) is determined by combining the
+            values of  ``days``, ``hours``, and ``minutes``, and
+            subtracting the resulting duration from the end time.
+
+            It is also allowed to omit the end time and duration here,
+            in which case
+            :meth:`~gcloud.monitoring.query.Query.select_interval`
+            must be called before the query is executed.
+
+        :type days: integer
+        :param days: The number of days in the time interval.
+
+        :type hours: integer
+        :param hours: The number of hours in the time interval.
+
+        :type minutes: integer
+        :param minutes: The number of minutes in the time interval.
+
+        :rtype: :class:`~gcloud.monitoring.query.Query`
+        :returns: The query object.
+
+        :raises: :exc:`ValueError` if ``end_time`` is specified but
+            ``days``, ``hours``, and ``minutes`` are all zero.
+            If you really want to specify a point in time, use
+            :meth:`~gcloud.monitoring.query.Query.select_interval`.
+
+        .. _supported metrics: https://cloud.google.com/monitoring/api/metrics
+        """
+        return Query(self, metric_type,
+                     end_time=end_time,
+                     days=days, hours=hours, minutes=minutes)
+
+    def fetch_metric_descriptor(self, metric_type):
+        """Look up a metric descriptor by type.
+
+        Example::
+
+            >>> METRIC = 'compute.googleapis.com/instance/cpu/utilization'
+            >>> print client.fetch_metric_descriptor(METRIC)
+
+        :type metric_type: string
+        :param metric_type: The metric type name.
+
+        :rtype: :class:`~gcloud.monitoring.metric.MetricDescriptor`
+        :returns: The metric descriptor instance.
+
+        :raises: :class:`gcloud.exceptions.NotFound`
+        """
+        return MetricDescriptor._fetch(self, metric_type)
+
+    def list_metric_descriptors(self, filter=None):
+        """List all metric descriptors for the project.
+
+        Example::
+
+            >>> for descriptor in client.list_metric_descriptors():
+            ...     print descriptor.type
+
+        :type filter: string or None
+        :param filter: An optional filter string describing the metric
+            descriptors to be returned. See the `filter documentation`_.
+
+        :rtype: list of :class:`~gcloud.monitoring.metric.MetricDescriptor`
+        :returns: A list of metric descriptor instances.
+
+        .. _filter documentation:
+            https://cloud.google.com/monitoring/api/v3/filters
+        """
+        # Allow "filter" as a parameter name: pylint: disable=redefined-builtin
```

Comment:
The _caller_ doesn't define anything by passing a keyword argument using a
built-in name. There's no bad behavior by the caller. It's only on the
implementation side that we have set up a dastardly trap for the would-be
gcloud hacker who attempts to use the filter built-in in this three-line,
one-statement function body.

On Tue, Apr 19, 2016 at 8:54 PM, Danny Hermes <notifications@github.com>
wrote:

> In gcloud/monitoring/client.py
> <https://github.com/GoogleCloudPlatform/gcloud-python/pull/1691#discussion_r60336814>
> :
>
> > +        Example::
> > +
> > +            >>> for descriptor in client.list_metric_descriptors():
> > +            ...     print descriptor.type
> > +
> > +        :type filter: string or None
> > +        :param filter: An optional filter string describing the metric
> > +            descriptors to be returned. See the `filter documentation`_.
> > +
> > +        :rtype: list of :class:`~gcloud.monitoring.metric.MetricDescriptor`
> > +        :returns: A list of metric descriptor instances.
> > +
> > +        .. _filter documentation:
> > +            https://cloud.google.com/monitoring/api/v3/filters
> > +        """
> > +        # Allow "filter" as a parameter name: pylint: disable=redefined-builtin
>
> Because it leads to unexpected bugs. The namespace cascade from locals()
> to globals() to __builtins__ gets pre-empted when builtin names are used
> as locals or globals.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-python/pull/1691/files/adb7e102bb158a442e473166b66d5919334cfdbd#r60336814>
>